### PR TITLE
snapm: address v0.6.0 feature/fix release blockers

### PR DESCRIPTION
**Pull request by bmr-cymru**
Resolves: #741
Resolves: #746
Resolves: #748
Resolves: #749
Resolves: #751
Resolves: #752
Resolves: #753
Resolves: #754
Resolves: #755
Resolves: #756
Resolves: #757
Resolves: #758
Resolves: #760
Resolves: #761
Resolves: #762
Resolves: #763
Resolves: #764
Resolves: #765
Resolves: #766
Resolves: #767
Resolves: #768
Resolves: #769
Resolves: #770
Resolves: #771

## Summary by CodeRabbit

* **New Features**
  * Diff CLI: diffstat (--stat), new "summary" output format and cache controls (--cache-mode, --cache-expires); two small CLI demos for diffing and progress.

* **Documentation**
  * Man page and user guide updated with formats, diffstat semantics, cache behaviour (default 900s) and cache option notes; system tmpfiles rule added for cache dir.

* **Bug Fixes**
  * Improved interrupt handling, rendering and tree/move handling.

* **Tests**
  * Expanded coverage for diffs, diffstat, binary/content diffs, caching and edge cases.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: tests/fsdiff/test_engine.py [hunk 1]
```diff
@@ -134,6 +134,112 @@ def test_diff_rendering_color(self):
         self.assertIn("<GREEN>", diff)
         self.assertIn("<WHITE>", diff)
 
+    def test_results_summary_with_diffstat(self):
+        summary = self.results.summary(diffstat=True, color="never")
+        self.assertIn("Total changes:     3", summary)
+        self.assertIn(" a |    2", summary)
```
[the remark this answers is not in the document]
Comment (bmr-cymru) [reply tests/fsdiff/test_engine.py]: @coderabbitai: same, it's a dupe. Just nix it.

File: scripts/difftest.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/python3
```
[the remark this answers is not in the document]
Comment (bmr-cymru) [reply scripts/difftest.py]: Fixed in commit b8b335e

File: scripts/difftest.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        "--file-types",
+        help="Generate file type information (slower)",
+        action="store_true",
+    )
+    parser.add_argument(
+        "-l",
+        "--log-level",
+        default = "info",
+        help = f"Set log level ({", ".join(LOG_LEVELS.keys())})",
+        choices=LOG_LEVELS.keys(),
```
[the remark this answers is not in the document]
Comment (bmr-cymru) [reply scripts/difftest.py]: Fixed in commit b8b335e

File: scripts/difftest.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    )
+    fsd = FsDiffer(m, options=options)
+    try:
+        results = fsd.compare_roots(snap_root, sys_root)
+    except (KeyboardInterrupt, SystemExit, BrokenPipeError):
+        return
+    print(f"Found {len(results)} changes:")
+    print("\n\n".join(str(record) for record in results))
+    if did_mount:
+        m.mounts.umount(snapset)
```
[the remark this answers is not in the document]
Comment (bmr-cymru) [reply scripts/difftest.py]: Fixed in commit b8b335e

File: tests/fsdiff/test_contentdiff.py [hunk 5]
```diff
@@ -84,6 +85,28 @@ def test_json_diff_structure(self):
...
+        old_path = self._create_file("old.bin", b"\x00", binary=True)
+        new_path = self._create_file("new.bin", b"\x01", binary=True)
+        info = FileTypeInfo("application/octet-stream", "Bin", FileTypeCategory.BINARY)
+        old_entry = make_entry(str(old_path), content_hash="h1")
+        new_entry = make_entry(str(new_path), content_hash="h2")
+        new_entry.file_type_info = info
+        diff = self.manager.generate_content_diff(old_path, new_path, old_entry, new_entry)
+        self.assertEqual(diff.diff_type, "binary")
+        self.assertTrue(diff.has_changes)
+
```
[the remark this answers is not in the document]
Comment (bmr-cymru) [reply tests/fsdiff/test_contentdiff.py]: Fixed in commit c7fc8e5

File: scripts/progtest.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+    "home",
+    "opt",
+    "srv",
+    "stuff",
+    "things",
+    "quux",
+    "foo",
+    "bar",
+    "baz",
+)
```
[the remark this answers is not in the document]
Comment (bmr-cymru) [reply scripts/progtest.py]: Fixed in commit 18c2582.

File: snapm/fsdiff/engine.py [hunk 7]
```diff
@@ -686,139 +824,152 @@ def compute_diff(
...
+
+            end_time = datetime.now()
+
+        except KeyboardInterrupt:
+            progress.cancel("Quit!")
+            raise
+        except SystemExit:
+            progress.cancel("Exiting.")
 
         progress.end(f"Found {len(diffs)} differences in {end_time - start_time}")
```
[the remark this answers is not in the document]
Comment (bmr-cymru) [reply snapm/fsdiff/engine.py]: Fixed in commit 854bf05.

File: tests/fsdiff/test_tree.py [hunk 8]
```diff
@@ -53,6 +53,13 @@ def test_init_root_none(self):
         with self.assertRaises(ValueError):
             DiffTree(None, 0)
 
+    def test_init_encoding_error(self):
+        with patch("sys.stdout") as mock_stdout:
+            mock_stdout.encoding = "ascii"
+            tc = TermControl(term_stream=mock_stdout)
+            tree = DiffTree(self.root, 0, term_control=tc)
+            self.assertEqual(tree.branch, "|-- ")
+
```
[the remark this answers is not in the document]
Comment (bmr-cymru) [reply tests/fsdiff/test_tree.py]: Fixed in commit 0fe5763.

File: tests/fsdiff/test_treewalk.py [hunk 9]
```diff
@@ -145,6 +163,12 @@ def test_FsEntry_to_dict(self):
         # Verify hex fallback
         self.assertEqual(data["xattrs"]["user.bin"], "ffff")
 
+    def test_FsEntry_to_dict_binary_xattr(self):
+        entry = make_entry("/file")
+        entry.xattrs = {"user.bin": b"\x80\x81"}
+        data = entry.to_dict()
+        self.assertEqual(data["xattrs"]["user.bin"], "8081")
+
```
[the remark this answers is not in the document]
Comment (bmr-cymru) [reply tests/fsdiff/test_treewalk.py]: Fixed in commit 0fe5763.

File: systemd/tmpfiles.d/snapm.conf [hunk 10]
```diff
@@ -2,6 +2,8 @@
 d /run/snapm 0700 root root - -
 d /run/snapm/mounts 0700 root root - -
 d /run/snapm/lock 0700 root root - -
+d /var/cache/snapm 0700 root root - -
 z /run/snapm - - - -
 z /run/snapm/mounts - - - -
 z /run/snapm/lock - - - -
+Z /var/cache/snapm - - - -
```
[the remark this answers is not in the document]
Comment (bmr-cymru) [reply systemd/tmpfiles.d/snapm.conf]: @coderabbitai: this case is different. The `/var/cache/snapm` directory is persistent and should be labelled properly. We want `tmpfiles` to recurse into `/var/cache/snapm/diffcache`.

**Comment by bmr-cymru** [comment]
@coderabbitai: re:
> Clarify the expiry logic branching.

Actually we can simplify further - if `expires == 0` we NEVER prune and `_valid()` always returns `True`, so the whole `prune` variable can go away and simply become:

```python
        valid = _valid(timestamp, expires=expires)
        cache_path = os.path.join(_DIFF_CACHE_DIR, file_name)
        if not valid:
            _log_info("Pruning expired cache file %s", file_name)
            try:
                os.unlink(cache_path)
            except OSError as err:
                _log_error("Error unlinking stale cache file %s: %s", file_name, err)
            continue
```

File: snapm/command.py [hunk 11]
```diff
@@ -92,6 +92,7 @@
 from .fsdiff import DiffOptions, FsDiffer, FsDiffResults
 
 DIFF_FORMATS = FsDiffResults.DIFF_FORMATS
+DIFF_CACHE_MODES = ["auto", "never", "always"]
```
[the remark this answers is not in the document]
Comment (bmr-cymru) [reply snapm/command.py]: Fixed in commit 2b29430

File: snapm/fsdiff/engine.py [hunk 12]
```diff
@@ -394,6 +416,65 @@ def _rstrip_nl(s: str) -> str:
...
+        return header + f"{added + removed:4} {plus * added}{minus * removed}"
+
+    path_width = max(len(record.path) for record in records) - 1
+    count = len(records)
+    diffstat = "\n".join(_render_one(record) for record in records)
+    trailer = (
+        f"\n {count} file{'s' if count > 1 else ''} changed, "
+        f"{adds} insertions(+), {dels} deletions(-)"
+    )
+    return diffstat + trailer
```
[the remark this answers is not in the document]
Comment (bmr-cymru) [reply snapm/fsdiff/engine.py]: Fixed in commit 4292fa2

File: tests/fsdiff/test_contentdiff.py [hunk 13]
```diff
@@ -193,6 +205,10 @@ def test_ContentDiff_serialization(self):
         self.assertEqual(data["old_content"], "a")
         self.assertEqual(data["diff_data"], ["-a", "+b"])
 
+    def test_differ_priority(self):
+        self.assertEqual(TextContentDiffer().priority, 10)
+        self.assertEqual(BinaryContentDiffer().priority, 5)
```
[the remark this answers is not in the document]
Comment (bmr-cymru) [reply tests/fsdiff/test_contentdiff.py]: @coderabbitai: are you high? There is only one `test_differ_priority()` as of [aced8e0](https://github.com/snapshotmanager/snapm/commit/aced8e0aac1da2f19fd01b5e779903d34531f8a1): the one at line 208. And it doesn't test `ContentDiff.to_dict()`, it tests differ priorities:

```python
    def test_differ_priority(self):
        self.assertEqual(TextContentDiffer().priority, 10)
        self.assertEqual(BinaryContentDiffer().priority, 5)
```

File: scripts/difftest.py [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+        metavar="PATH",
+        dest="from_path",
+        help="Start traversal from PATH",
+    )
+    parser.add_argument(
+        "diff_from",
+        type=str,
+        help="Snapshot set name to compare to system",
+    )
+    args = parser.parse_args()
```
[the remark this answers is not in the document]
Comment (bmr-cymru) [reply scripts/difftest.py]: Fixed in commit 780cdfd.
